src/topology/homotopy/product.lean

```diff
@@ -0,0 +1,83 @@
+/-
+Copyright (c) 2021 Praneeth Kolichala. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Praneeth Kolichala
+-/
+import topology.homotopy.basic
+import topology.constructions
+
+/-!
+# Product of homotopies
+
+In this file, we introduce definitions for the product of
+homotopies. We show that the products of relative homotopies
+are still relative homotopies.
+
+## Definitions
+- `homotopy.pi homotopies`: Let f and g be a family of functions
+  indexed on I, such that for each i ∈ I, fᵢ and gᵢ are maps from A to Xᵢ.
+  Let `homotopies` be a family of homotopies from fᵢ to gᵢ for each i.
+  Then `homotopy.pi homotopies` is the canonical homotopy
+  from ∏ f to ∏ g, where ∏ f is the product map from A to Πi, Xᵢ,
+  and similarly for ∏ g. All homotopies are done relative to some set S ⊆ A.
+
+- `homotopy.prod F G` is the product of homotopies F and G,
+   where F is a homotopy between f₀ and f₁, G is a homotopy between g₀ and g₁.
+   The result F × G is a homotopy between (f₀ × g₀) and (f₁ × g₁).
+   Again, all homotopies are done relative to S.
+-/
+
+noncomputable theory
+
+namespace continuous_map.homotopy
+open continuous_map
+
+section pi
+
+variables {I : Type*} {X : I → Type*} [∀i, topological_space (X i)]
+  {A : Type*} [topological_space A]
+  {f g : Π i, C(A, X i)} {S : set A}
+
+
+/-- The product homotopy of `homotopies` between functions `f` and `g` -/
+def pi (homotopies : Π i : I, homotopy_rel (f i) (g i) S) :
+  homotopy_rel (pi f) (pi g) S :=
```

Comment:
I'm fairly sure this holds for `continuous_map.homotopy` as well? In which case I would do this for general homotopies and not just relative to a set. You can then use `continuous_map.homotopy.pi` to define `continuous_map.homotopy_rel.pi`, and that should only require you to prove the `prop'` field, as the others should work as is.

```lean
def pi (homotopies : Π i : I, homotopy (f i) (g i)) :
  homotopy (pi f) (pi g) :=
```

Reply:
Technically (i.e. mathematically), it's `homotopy_rel` which is the more general case, since `homotopy` is `homotopy_rel` specialized with `S = {}`, which is why I originally only proved `homotopy_rel`. But programming-wise you're right. I added non-relative versions.